torch/_subclasses/meta_utils.py

```diff
@@ -9,36 +11,133 @@ def safe_is_leaf(t):
 # share storage because this is how we correlate shared storages to the same
-# meta storages; similarly, it's important NOT to use the same object for
-# unrelated groups of tensors because this class will remember all the
-# tensors/storages its seen and therefore leak memory.
+# meta storages. This class will hold weak references to cached tenosrs
+# and tensor storages.
 class MetaConverter:
     def __init__(self):
         self.storage_memo = {}
         self.tensor_memo = {}
+        self.maybe_storages_to_delete = []
+        self.check_expired_frequency = 128
+        self.check_expired_count = 0
         self.hit = 0
         self.miss = 0
+        self.del_hook = None
 
     def successful(self):
         return self.hit > 0 and self.miss == 0
 
+    def check_for_expired_weak_storages(self):
+        new_li = []
+        stor_to_delete = []
+        for obj in self.maybe_storages_to_delete:
+            if not obj.expired():
+                new_li.append(obj)
+            else:
+                stor_to_delete.append(obj)
+        for obj in stor_to_delete:
+            self.storage_memo.pop(obj, None)
+        self.maybe_storages_to_delete = new_li
+
+        # if for some reason we have aquired many storages which have not expired
+        # even though a tensor with their storage has expired (aliasing or otherwise)
+        # check for expired storages less often so as to bound the amount of work we
+        # do checking for expired storages
+        self.check_expired_frequency = max(self.check_expired_frequency, len(self.maybe_storages_to_delete))
+
+
+    def get_tensor_memo(self, t):
+        return self.tensor_memo.get(WeakTensorRefKey(t), None)
+
+    def set_tensor_memo(self, t, v):
+        # hold a weak ref to self, otherwise it will be kept alive
+        # by the del_ten closure
+        self_weak_ref = weakref.ref(self)
+        weak_st = StorageWeakRef(t.storage())
+        tensor_ref_key = WeakTensorRefKey(t)
+
+        def del_ten():
+            # tensor outlives the converter
+            self_ref = self_weak_ref()
+            if self_ref is None:
+                return
+            # on shutdown, tensor_ref_key may not be in memo
+            self_ref.tensor_memo.pop(tensor_ref_key, None)
+            if weak_st and weak_st.expired():
+                self_ref.storage_memo.pop(weak_st, None)
+            else:
+                # [expired-storages]
+                # NB: even though the tensor has died,
+                # the deallocation of its storage can take longer,
+                # even when the storage has no other uses/views.
+                # In this case, the StorageWeakRef object will be kept alive
+                # longer than it needs to be, however the storage itself
+                # will be deallocated. We retain the possibly dead storages
+                # and periodically check if any of them are expired and
+                # can be freed.
```

Comment:
Wouldn't a more conventional design be to attach finalizer to the storage, and maintain storage memo removal separately?

Reply:
As of now at least, there's no binding of Storage in python that would be deallocated when it's corresponding tensors are deallocated, as far as I'm aware of. that would be nice though..